src/jwt.ts

```diff
@@ -53,37 +60,136 @@ export const decode = (token: string) => {
-    throw new Error('Invalid issuer');
+    throw new Error(
+      `Issuer (iss) claim mismatch in the ID token; expected "${options.iss}", found "${decoded.claims.iss}"`
+    );
   }
-  if (decoded.claims.aud !== options.aud) {
-    throw new Error('Invalid audience');
+
+  if (!decoded.user.sub) {
+    throw new Error(
+      'Subject (sub) claim must be a string present in the ID token'
+    );
   }
+
   if (decoded.header.alg !== 'RS256') {
-    throw new Error('Invalid algorithm');
+    throw new Error(
+      `Signature algorithm of "${decoded.header.alg}" is not supported. Expected the ID token to be signed with "RS256".`
+    );
+  }
+
+  if (
+    !decoded.claims.aud ||
+    !(
+      typeof decoded.claims.aud === 'string' ||
+      Array.isArray(decoded.claims.aud)
+    )
+  ) {
+    throw new Error(
+      'Audience (aud) claim must be a string or array of strings present in the ID token'
+    );
   }
-  if (decoded.claims.nonce !== options.nonce) {
-    throw new Error('Invalid nonce');
+  if (Array.isArray(decoded.claims.aud)) {
+    if (!decoded.claims.aud.includes(options.aud)) {
+      throw new Error(
+        `Audience (aud) claim mismatch in the ID token; expected "${
+          options.aud
+        }" but was not one of "${decoded.claims.aud.join(', ')}"`
+      );
+    }
+    if (decoded.claims.aud.length > 1) {
+      if (!decoded.claims.azp) {
+        throw new Error(
+          'Authorized Party (azp) claim must be a string present in the ID token when Audience (aud) claim has multiple values'
+        );
+      }
+      if (decoded.claims.azp !== options.aud) {
+        throw new Error(
+          `Authorized Party (azp) claim mismatch in the ID token; expected "${options.aud}", found "${decoded.claims.azp}"`
+        );
+      }
+    }
+  } else if (decoded.claims.aud !== options.aud) {
+    throw new Error(
+      `Audience (aud) claim mismatch in the ID token; expected "${options.aud}" but found "${decoded.claims.aud}"`
+    );
+  }
+  if (options.nonce) {
+    if (!decoded.claims.nonce) {
+      throw new Error(
+        'Nonce (nonce) claim must be a string present in the ID token'
+      );
+    }
+    if (decoded.claims.nonce !== options.nonce) {
+      throw new Error(
+        `Nonce (nonce) claim mismatch in the ID token; expected "${options.nonce}", found "${decoded.claims.nonce}"`
+      );
+    }
+  }
+
+  if (options.max_age && !isNumber(decoded.claims.auth_time)) {
+    throw new Error(
+      'Authentication Time (auth_time) claim must be a number present in the ID token when Max Age (max_age) is specified'
+    );
+  }
+
+  /* istanbul ignore next */
+  if (!isNumber(decoded.claims.exp)) {
+    throw new Error(
+      'Expiration Time (exp) claim must be a number present in the ID token'
+    );
+  }
+  if (!isNumber(decoded.claims.iat)) {
+    throw new Error(
+      'Issued At (iat) claim must be a number present in the ID token'
+    );
   }
 
+  const leeway = options.leeway || 60;
```

Comment:
why do we set default leeway to 60 seconds? There is no mention about it in our docs.

Reply:
We decided to set the default leeway to 60 across the board to keep the consistency.

Reply:
Makes sense. Let's include this piece of information here: https://auth0.github.io/auth0-spa-js/interfaces/auth0clientoptions.html#leeway